pxtblocks/layout.ts

```diff
@@ -384,65 +384,96 @@ export function documentToSvg(xsg: Node): string {
+    const dataUriPromises = images
         .filter(image => {
             const href = image.getAttributeNS(XLINK_NAMESPACE, "href");
             return href && !/^data:/.test(href);
         })
-        .map(img => img as unknown as HTMLImageElement)
-        .map((image: HTMLImageElement) => {
+        .map(async image => {
             const href = image.getAttributeNS(XLINK_NAMESPACE, "href");
             let dataUri = imageXLinkCache[href];
-            return (dataUri ? Promise.resolve(imageXLinkCache[href])
-                : pxt.BrowserUtils.loadImageAsync(image.getAttributeNS(XLINK_NAMESPACE, "href"))
-                    .then((img: HTMLImageElement) => {
-                        const cvs = document.createElement("canvas") as HTMLCanvasElement;
-                        const ctx = cvs.getContext("2d");
-                        let w = img.width
-                        let h = img.height
-                        cvs.width = w;
-                        cvs.height = h;
-                        ctx.drawImage(img, 0, 0, w, h, 0, 0, cvs.width, cvs.height);
-                        imageXLinkCache[href] = dataUri = cvs.toDataURL("image/png");
-                        return dataUri;
-                    }).catch(e => {
-                        // ignore load error
-                        pxt.debug(`svg render: failed to load ${href}`);
-                        return "";
-                    }))
-                .then(href => { image.setAttributeNS(XLINK_NAMESPACE, "href", href); })
+
+            if (!dataUri) {
+                try {
+                    const img = await pxt.BrowserUtils.loadImageAsync(image.getAttributeNS(XLINK_NAMESPACE, "href"));
+                    const cvs = document.createElement("canvas") as HTMLCanvasElement;
+                    const ctx = cvs.getContext("2d");
+                    let w = img.width
+                    let h = img.height
+                    cvs.width = w;
+                    cvs.height = h;
+                    ctx.drawImage(img, 0, 0, w, h, 0, 0, cvs.width, cvs.height);
+                    dataUri = cvs.toDataURL("image/png");
+                    imageXLinkCache[href] = dataUri;
+                }
+                catch (e) {
+                    // ignore load error
+                    pxt.debug(`svg render: failed to load ${href}`);
+                    dataUri = "";
+                }
+            }
+            image.setAttributeNS(XLINK_NAMESPACE, "href", href);
         });
-    return Promise.all(p).then(() => { })
+
+    const linkedIconPromises = images
+        .filter(image => /^\/.*.svg$/.test(image.getAttribute("href")))
+        .map(async image => {
+            const svgUri = image.getAttribute("href");
+
+            // Don't love hard coding these icon values, but the comment icons set their
+            // width/height using CSS so there isn't a great means of detecting it
+            let width = 24;
```

Comment:
would `window.getComputedStyle(image).getPropertyValue("width")` work? Not a big deal to hardcode this really / computed style has it's own downsides, but mainly curious.

Reply:
hmm does that work with svg elements? i'll give it a shot!

Reply:
does not work, not sure if it's because it's svg or because it's not in the document